sdk/keyvault/keyvault-keys/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -82,94 +101,82 @@ export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
+      response = await this._nextPolicy.sendRequest(webResource);
+      webResource.body = originalBody;
     } else {
-      // or use the cached token if we have one
-      await this.authenticateRequest(webResource);
+      // If we did have a challenge in memory,
+      // we attempt to load the token from the cache into the request before we try to send the request.
+      await this.loadToken(webResource);
+      response = await this._nextPolicy.sendRequest(webResource);
     }
 
-    const response = await this._nextPolicy.sendRequest(webResource);
-
-    if (response.status == 401) {
-      webResource.body = originalBody;
-
-      let www_authenticate = response.headers.get("WWW-Authenticate");
-
-      if (www_authenticate) {
-        // The challenge based authentication will contain both an authorization URI with a token,
-        // and the resource to which that token is valid against (also called the scope).
-        const { authorization, resource } = this.parseWWWAuthenticate(www_authenticate);
-        const challenge = new AuthenticationChallenge(authorization, resource + "/.default")
-
-        if (!challenge.equalTo(this.challengeCache.challenge)) {
-          this.challengeCache.setCachedChallenge(challenge);
-          this.tokenCache.setCachedToken(undefined);
-
-          await this.authenticateRequest(webResource);
-          return this._nextPolicy.sendRequest(webResource);
-        }
-        return response;
-      }
+    // If we don't receive a response with a 401 status code,
+    // then we can assume this response has nothing to do with the challenge authentication process.
+    if (response.status !== 401) {
       return response;
-    } else {
+    }
+
+    // If the response status is 401, we only re-authenticate if the WWW-Authenticate header is present.
+    const wwwAuthenticate = response.headers.get("WWW-Authenticate");
+    if (!wwwAuthenticate) {
       return response;
     }
+
+    // The challenge based authentication will contain both:
+    // - An authorization URI with a token,
+    // - The resource to which that token is valid against (also called the scope).
+    const parsedWWWAuth = this.parseWWWAuthenticate(wwwAuthenticate);
+    const authorization = parsedWWWAuth.authorization! || parsedWWWAuth.authorization_url!;
+    const resource = parsedWWWAuth.resource! || parsedWWWAuth.scope!;
+
+    const challenge = new AuthenticationChallenge(authorization, resource + "/.default");
+
+    // Either if there's no cached challenge at this point (could have happen in parallel),
+    // or if the cached challenge has a different scope,
+    // we store the just received challenge and reset the cached token, to force a re-authentication.
+    // This is exactly what C# is doing, as we can see here:
+    // https://github.com/heaths/azure-sdk-for-net/blob/a35434c1ce955d4cdeb393748f9f6407d4a984e2/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L233
```

Comment:
I appreciate the nod, but might want to link to the main repo's source? 😉

Reply:
I mean, I would get lost in the main repo's source! I want to link to what's similar in that code. Let me know if you know a better link.

Reply:
*Maybe* what you want is this?

https://github.com/Azure/azure-sdk-for-net/blob/70e54b878ff1d01a45266fb3674a396b4ab9c1d2/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L143-L147

Reply:
Wait, so you do compare the authority? hmm

Reply:
Ok! Updated!